rootfs/etc/nginx/lua/test/balancer_test.lua

```diff
@@ -325,8 +331,86 @@ describe("Balancer", function()
         end)
       end)
 
+    describe("canary by query", function()
+      it("returns correct result for given querys", function()
+        local test_patterns = {
+          -- with no query value setting
+          {
+            case_title = "no custom query value and query value is 'always'",
+            query_name = "canaryQuery",
+            query_value = "",
+            request_query_name = "canaryQuery",
+            request_query_value = "always",
+            expected_result = true,
+          },
+          {
+            case_title = "no custom query value and query value is 'never'",
+            query_name = "canaryQuery",
+            query_value = "",
+            request_query_name = "canaryQuery",
+            request_query_value = "never",
+            expected_result = false,
+          },
+          {
+            case_title = "no custom query value and query value is undefined",
+            query_name = "canaryQuery",
+            query_value = "",
+            request_query_name = "canaryQuery",
+            request_query_value = "foo",
+            expected_result = false,
+          },
+          {
+            case_title = "no custom query value and query name is undefined",
+            query_name = "canaryQuery",
+            query_value = "",
+            request_query_name = "foo",
+            request_query_value = "always",
+            expected_result = false,
+          },
+          -- with query value setting
+          {
+            case_title = "custom query value is set and query value is 'always'",
+            query_name = "canaryQuery",
+            query_value = "foo",
+            request_query_name = "canaryQuery",
+            request_query_value = "always",
+            expected_result = false,
+          },
+          {
+            case_title = "custom query value is set and query value match custom query value",
+            query_name = "canaryQuery",
+            query_value = "foo",
+            request_query_name = "canaryQuery",
+            request_query_value = "foo",
+            expected_result = true,
+          },
+          {
+            case_title = "custom query value is set and query name is undefined",
+            query_name = "canaryQuery",
+            query_value = "foo",
+            request_query_name = "bar",
+            request_query_value = "foo",
+            expected_result = false
+          },
+        }
```

Comment:
A test scenario for the regex feature will also be nice. You may also want a test that confirms the higher precedence by value over the regex as described in the doc update.

Reply:
Maybe I spoke too soon, I see this is covered in the ginkgo tests and I’m guessing they aren’t in Lua tests because we can’t mock `ngx.re` behavior ?

Reply:
I tested the same test scenario that has been done for the canary-by-header annotations.
I was wondering also why the regex tests are only tested in the e2e tests.